pkg/secretless/config/v1/config.go

```diff
@@ -0,0 +1,163 @@
+package v1
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/go-ozzo/ozzo-validation"
+
+	"gopkg.in/yaml.v2"
+)
+
+// StoredSecret represents not the value of a "secret," but the abstract concept
+// of "a secret stored somewhere".
+//
+// Note that "Name" will be the key that maps to this secret's actual value in
+// the map[string][]byte when the "StoredSecret" itself is looked up by a
+// Resolver.
+type StoredSecret struct {
+	// How client code will refer to the secret's actual value at runtime.
+	// Specifically, the key to the secret's value in the map[string][]byte
+	// returned by a Resolver.
+	Name string
+	Provider string
+	// The identifier within the context of a Provider.  Ie, how a provider
+	// refers to this secret.  Eg, a database primary key.
+	ID string
+}
+
+// Listener listens on a port on socket for inbound connections, which are
+// handed off to Handlers.
+type Listener struct {
+	Address     string
+	CACertFiles []string `yaml:"caCertFiles"`
+	Debug       bool
+	Name        string
+	Protocol    string
+	Socket      string
+}
+
+// Handler processes an inbound message and connects to a specified backend
+// using Credentials which it fetches from a provider.
+type Handler struct {
+	Name         string
+	Type         string
+	ListenerName string `yaml:"listener"`
+	Debug        bool
+	Match        []string         `yaml:"match"`
+	Patterns     []*regexp.Regexp `yaml:"-"`
+	Credentials  []StoredSecret
+}
+
+// Config is the main configuration structure for Secretless.
+// It lists and configures the protocol listeners and handlers.
+type Config struct {
+	Listeners []Listener
+	Handlers  []Handler
+}
+
+// HasCredential indicates whether a Handler has the specified credential.
+func (h Handler) HasCredential(credentialName string) bool {
+	for _, c := range h.Credentials {
```

Comment:
I know it's old code, but `c` -> `credential`. Up to you if you want to include that as part of the PR.

Reply:
done